src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -1575,18 +1577,17 @@ private TypeDeclarationSyntax ParseClassOrStructOrInterfaceOrExtensionDeclaratio
             var name = this.ParseIdentifierToken();
             var typeParameters = this.ParseTypeParameterList();
 
-            var paramList = CurrentToken.Kind == SyntaxKind.OpenParenToken && mainKeyword.Kind != SyntaxKind.ExtensionKeyword // PROTOTYPE
-                ? ParseParenthesizedParameterList() : null;
+            var paramList = CurrentToken.Kind == SyntaxKind.OpenParenToken
+                ? ParseParenthesizedParameterList()
+                : null;
 
-            SyntaxToken? forKeyword = null;
-            TypeSyntax? forType = null;
-            if (mainKeyword.Kind == SyntaxKind.ExtensionKeyword
-                && CurrentToken.Kind == SyntaxKind.ForKeyword)
-            {
-                // PROTOTYPE consider error recovery for `class X for type`
-                forKeyword = EatToken(SyntaxKind.ForKeyword);
-                forType = ParseType();
-            }
+            if (paramList != null && mainKeyword.Kind == SyntaxKind.ExtensionKeyword)
+                paramList = AddError(paramList, ErrorCode.ERR_ExtensionPrimaryConstructor);
+
+            // PROTOTYPE. Parse this for all type declarations and give good error message.
```

Comment:
i'm willing to do this @jcouv .  will do in a followup though.

Reply:
Thanks much!
There's another PROTOTYPE comment for improving parsing below, if you have time: "// PROTOTYPE consider improving error recovery for `extension explicit` and `extension implicit`" (wrong order)

Reply:
yup.  happy to take that.